cleo/commands/completions_command.py

```diff
@@ -120,266 +120,148 @@ def handle(self) -> int:
-
-        commands = []
-        global_options = set()
-        options_descriptions = {}
-        commands_options = {}
-        for option in self.application.definition.options:
-            options_descriptions[
-                "--" + option.name
-            ] = self.io.output.formatter.remove_format(option.description)
-            global_options.add("--" + option.name)
-
-        for command in self.application.all().values():
-            if not command.enabled or command.hidden:
-                continue
-
-            command_options = []
-            commands.append(command.name)
 
-            options = command.definition.options
-            for option in sorted(options, key=lambda o: o.name):
-                name = "--" + option.name
-                description = option.description
-                command_options.append(name)
-                options_descriptions[name] = description
+        return script_name, script_path
 
-            commands_options[command.name] = command_options
-
-        compdefs = "\n".join(
-            [f"complete -o default -F {function} {alias}" for alias in aliases]
-        )
-
-        commands = sorted(commands)
+    def render_bash(self) -> str:
+        script_name, script_path = self._get_script_name_and_path()
+        aliases = [script_name, script_path, *self.option("alias")]
+        function = self._generate_function_name(script_name, script_path)
 
-        command_list = []
-        for i, command in enumerate(commands):
-            options = set(commands_options[command]).difference(global_options)
-            options = sorted(options)
-            options = [self._zsh_describe(opt, None).strip('"') for opt in options]
+        # Global options
+        assert self.application
+        opts = [
+            f"--{opt.name}"
+            for opt in sorted(self.application.definition.options, key=lambda o: o.name)
+        ]
 
-            desc = [
-                f"            ({command})",
-                f'            opts="${{opts}} {" ".join(options)}"',
+        # Commands + options
+        cmds = []
+        cmds_opts = []
+        for cmd in sorted(self.application.all().values(), key=lambda c: c.name or ""):
+            if cmd.hidden or not cmd.enabled or not cmd.name:
+                continue
+            cmds.append(cmd.name)
+            options = " ".join(
+                f"--{opt.name}".replace(":", "\\:")
+                for opt in sorted(cmd.definition.options, key=lambda o: o.name)
+            )
+            cmds_opts += [
+                f"            ({cmd.name})",
+                f'            opts="${{opts}} {options}"',
                 "            ;;",
+                "",  # newline
             ]
 
-            if i < len(commands) - 1:
-                desc.append("")
-
-            command_list.append("\n".join(desc))
-
-        output = template % {
+        return TEMPLATES["bash"] % {
```

Comment:
While you are on it, you could change this modulo formatting to `.format()`. (here and in other methods)

Reply:
I had the same thought until I realized that both bash and zsh templates contain literal brackets (like `${cur}`), so the extra escaping required made it awkward

Reply:
Oh, right, ok. Nevermind then.